// @reown/appkit-siwx/dist/esm/tests/ui/utils/HelpersUtil.js
export class HelpersUtil {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static getByTestId(element, testId) {
        return element.shadowRoot?.querySelector(`[data-testid="${testId}"]`) || null;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static queryByTestId(element, testId) {
        return element.shadowRoot?.querySelector(`[data-testid="${testId}"]`) || null;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static getAllByTestId(element, testId) {
        return Array.from(element.shadowRoot?.querySelectorAll(`[data-testid="${testId}"]`) || []);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static querySelector(element, selector) {
        return element.shadowRoot?.querySelector(selector) || null;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static querySelectorAll(element, selector) {
        return Array.from(element.shadowRoot?.querySelectorAll(selector) || []);
    }
}
